decompiler/util/data_decompile.cpp

```diff
@@ -578,6 +590,147 @@ goos::Object sp_launch_grp_launcher_decompile(const std::vector<LinkedWord>& wor
                                                file, TypeSpec("sparticle-group-item"), 32);
 }
 
+goos::Object decompile_sound_spec(const TypeSpec& type,
+                                  const DecompilerLabel& label,
+                                  const std::vector<DecompilerLabel>& labels,
+                                  const std::vector<std::vector<LinkedWord>>& words,
+                                  const TypeSystem& ts,
+                                  const LinkedObjectFile* file) {
+  // auto normal = decompile_structure(type, label, labels, words, ts, file, false);
+  // fmt::print("Doing: {}\n", normal.print());
+  auto uncast_type_info = ts.lookup_type(type);
+  auto type_info = dynamic_cast<StructureType*>(uncast_type_info);
+  if (!type_info) {
+    throw std::runtime_error(fmt::format("Type {} wasn't a structure type.", type.print()));
+  }
+  ASSERT(type_info->get_size_in_memory() == 0x4c);
+
+  // get words for real
+  auto offset_location = label.offset - type_info->get_offset();
+  int word_count = (type_info->get_size_in_memory() + 3) / 4;
+  std::vector<LinkedWord> obj_words;
+  obj_words.insert(obj_words.begin(),
+                   words.at(label.target_segment).begin() + (offset_location / 4) + 1,
+                   words.at(label.target_segment).begin() + (offset_location / 4) + word_count);
+
+  for (int i = 0; i < word_count - 1; ++i) {
+    if (i == word_count - 2 && !obj_words.at(i).data) {
+      // just some default initialized sound spec, don't attempt anything fancy.
+      return decompile_structure(type, label, labels, words, ts, file, false);
+    }
+    if (obj_words.at(i).data)
+      break;
+  }
+
+  u16 implicit_mask = 0;
+  u16 mask = obj_words.at(0).data;
+  float num = word_as_float(obj_words.at(1));
+  u8 group = obj_words.at(2).data;
+  char sound_name[17];
+  sound_name[16] = 0;
+  memcpy(&sound_name[0], &obj_words.at(3).data, sizeof(u32));
+  memcpy(&sound_name[4], &obj_words.at(4).data, sizeof(u32));
+  memcpy(&sound_name[8], &obj_words.at(5).data, sizeof(u32));
+  memcpy(&sound_name[12], &obj_words.at(6).data, sizeof(u32));
+  std::string name(sound_name);
+
+  for (int i = 0; i < 4; ++i) {
+    if (obj_words.at(7 + i).data) {
+      throw std::runtime_error("static sound-spec trans was not zero.");
+    }
+  }
+
+  s32 volume = word_as_s32(obj_words.at(11));
+  s32 pitch = word_as_s32(obj_words.at(12));
+  s32 bend = word_as_s32(obj_words.at(13));
+  s16 fo_min = word_as_s32(obj_words.at(14));
+  s16 fo_max = word_as_s32(obj_words.at(14)) >> 16;
+  s8 fo_curve = word_as_s32(obj_words.at(15));
+  s8 priority = word_as_s32(obj_words.at(15)) >> 8;
+  s32 auto_time = word_as_s32(obj_words.at(16));
+  s32 auto_from = word_as_s32(obj_words.at(17));
```

Comment:
it might be possible to reuse some of the level unpacking code here, if you end up doing a lot more of these. You can do stuff like `read_plain_data_field<s16>(ref, "fo_min", dts)`.